dartagnan/src/main/java/com/dat3m/dartagnan/wmm/relation/base/memory/RelCo.java

```diff
@@ -215,12 +218,18 @@ protected BooleanFormula encodeApprox(SolverContext ctx) {
                     }
 
                     IExpr address = init.getAddress();
-                    IntegerFormula a1 = convertToIntegerFormula(w1.getMemAddressExpr(), ctx);
-                    IntegerFormula a2 = convertToIntegerFormula(address.toIntFormula(init,ctx), ctx);
-                    IntegerFormula v1 = convertToIntegerFormula(w1.getMemValueExpr(), ctx);
-                    IntegerFormula v2 = convertToIntegerFormula(init.getBase().getLastMemValueExpr(ctx,init.getOffset()), ctx);
-                    BooleanFormula sameAddress = imgr.equal(a1, a2);
-                    BooleanFormula sameValue = imgr.equal(v1, v2);
+                    Formula a1 = w1.getMemAddressExpr();
+                    Formula a2 = address.toIntFormula(init,ctx);
+                    Formula v1 = w1.getMemValueExpr();
+                    Formula v2 = init.getBase().getLastMemValueExpr(ctx,init.getOffset());
+                    // The boogie file might have a different type (Ints vs BVs) that the imposed by ARCH_PRECISION
+                    // In such cases we perform the transformation 
+                    BooleanFormula sameAddress = ARCH_PRECISION > -1 ? 
+                    		bvmgr.equal(convertToBitvectorFormula(a1, ctx), convertToBitvectorFormula(a2, ctx)) : 
+                        	imgr.equal(convertToIntegerFormula(a1, ctx), convertToIntegerFormula(a2, ctx));
+                    BooleanFormula sameValue = ARCH_PRECISION > -1 ? 
+                    		bvmgr.equal(convertToBitvectorFormula(v1, ctx), convertToBitvectorFormula(v2, ctx)) : 
+                        	imgr.equal(convertToIntegerFormula(v1, ctx), convertToIntegerFormula(v2, ctx));
```

Comment:
Again, `generalEquals` should do the trick

Reply:
if `ARCH_PRECISION == -1`, addresses will have this precision and thus when encoded they will be `IntegerFormula`s. If the boogie file was generated by telling smack to use bit-vectors, it is possible that the accessed address of a `MemEvent` has `precision > 0` (e.g. in the presence of pointers) and thus when encoded they will be `BitvectorFormuala`s. Thus the conversions are required.

I just discovered this because the svcomp-loop benchmarks (which were generated by tell smack to use bv) were failing. It seems there were no problems related to `Init` events and then I just missed the change there (but the change should also be applied there).

Reply:
I'm not sure I can follow. As far as I see you do have conversion operators when comparing to `Init` events (for the sake of defining `lastCo`) but you do not have those conversions when dealing with normal `Writes` (see the above comment on that). But according to what you say, exactly there you should run into problems, no?

Reply:
Sorry, it was the other way round. In summary, the same code (using conversions) shall be used everywhere. I think the best is to do the globalFal checking + conversions in the definition of general equal and use this method everywhere